### DISCUSSION TITLE
Bad code or hardware limitation - 2 city hubs with a train motor on each, bluetooth connected

### DISCUSSION THREAD

**Discussion by EmperorKabir** [pybricks/support#1398] [Q&A] [answered]
Hi,

I have used block coding (not confident with script) to get 2 trains with a hub and motor in each of the furthest ends (train driver cab) connected together to run according to remote control inputs. If left plus/minus is pressed, the power level increases or decreases respectively by 10%. If the right plus/minus is pressed, the train stays at 50% or -50% respectively while the button is held down. The red buttons are kill switches and the lights stay on permanently.

I have noticed that the 'support' hub, the one with the much smaller amount of code as shown below, seems to stutter and lag. Could anyone please confirm whether this is just a hardware limitation due to bluetooth (rather than being wired or on the same hub) or whether I can tweak my code, presumably the 'wait' blocks, so that this stutter does not occur? If it is the wait blocks, I've added these in to ensure that by pressing the buttons on the left, it doesnt register multiple inputs too quickly and increases by a multiple of 10 accidentally on just one press.

Main hub code for forward facing train:

![image](https://github.com/pybricks/support/assets/114411808/8de89b38-386a-4307-872d-f3613ad4b625)

Support hub code at back of train:

![image](https://github.com/pybricks/support/assets/114411808/4046ebb9-64db-4027-a319-9b0218047fe8)

Video example of stutter, (I've also off camera tested this by holding the trains in my hands and seeing the motor stop and start or move with a noticeable lag of around 0.5 seconds or so behind the main motor:

https://youtu.be/u9KsuxnKBfk

Edit:

Here's another (probably better) video, (still needs a few minutes of processing at the exact time of posting):

https://youtu.be/HhUiduABsGo

and I also tried adding a 20ms wait to the end of the support hub loop but it seemed to make no difference.

**Comment by laurensvalk** [comment]
Are the hubs connected to the computer at the time of recording? The hubs communicate a bit better when they aren't also connected to the computer at the same time.

> seeing the motor stop and start or move with a noticeable lag of around 0.5 seconds or so behind the main motor:

Foreseeing this, here is [example code to power the motors from the same hub](https://github.com/orgs/pybricks/discussions/1386#discussioncomment-8170882) :smile:

**Reply by laurensvalk** [reply]
Unrelated to this, but you may find it helpful to expand the if/else blocks as follows, so you don't need to nest them:

The end result is the same in your case, but it can help make it easier to follow.

![image](https://github.com/pybricks/support/assets/12326241/61d955c1-a01a-4457-abbc-55788540a735)

**Reply by EmperorKabir** [reply]
Haha yes I'm seeing how this could be beneficial but in this case with the length of the train and the hubs fitting in the endmost carriages, wiring them is going to be very difficult. 

The hubs are not connected to the computer no.

(I just realised how untidy my if blocks are woooops)

Is this then a hardware limitation and the code cannot be altered to reduce the lag/drops?

**Reply by laurensvalk** [reply]
I think it should be feasible to improve. I'd have to dig a bit more into your code to see.

For things like this, it usually helps to try it out in a small, separate program, then get back to the big one when you've figured it out.

For example, leave your "receiver hub" as is, but make an alternate "sender hub" program such as:

Repeat forever:
- Send 50%
- wait 10 seconds
- Send 0%
- wait 10 seconds.

Then see how smooth (or not) this synchronizes the trains. This could then be the benchmark to achieve in your bigger program.

**Reply by EmperorKabir** [reply]
Thank you, so it sounds like the most likely element that potentially could be improved is the use of the 'wait' blocks like I suspected?

**Comment by EmperorKabir** [comment]
I tried changing batteries and a simpler code but I ended up with the same problems and cannot seem to fix it. Sometimes the support one stops for a good 5 or so seconds before starting up again. I tried switching the code between the hubs and same problem in reverse. So it's a code issue or a hardware limitation but I decided that given I am simply wanting the hubs to do the opposite to each other, I would try and instead get them to connect to one remote. I cannot get this to work however. The first hub will connect but then the second one times out trying to find the remote:

Here is my code, (should I open a new discussion for this?):

![image](https://github.com/pybricks/support/assets/114411808/ee4b5c34-b2f0-4955-b0ef-d5b9a3fcc36e)

and the other hub:
![image](https://github.com/pybricks/support/assets/114411808/90e9b219-0b36-4088-b11f-d259373b8ab3)

**Reply by laurensvalk** [reply]
One remote can connect to only one hub. (That's a limitation in the remote, and we LEGO didn't provide a way to update it).

See an alternative program below :slightly_smiling_face:

**Comment by laurensvalk** [comment]
I tried out a minimalist program with two hubs and the remote. There is a slight delay as expected, but not that much.

It's not in the range of 5 seconds, so that aspect may be more about the program.

Put another way, can you describe in a few bullet points what you'd like your ideal program to do? Happy to help. I think it could be a great example project for others to try out too.

https://github.com/pybricks/support/assets/12326241/bf5b23cb-c89f-4d32-9ae3-08fca7a183db

![image](https://github.com/pybricks/support/assets/12326241/7fd9d0cb-c213-4d37-8aba-6280f3797268)

**Reply by EmperorKabir** [reply]
No problem and thank you again for your help!

My objective:

I have a lego train with a city hub connected to a motor and lights in the driver's carriage. I have bought 2 lego sets like this and stuck them back to back as can be seen in the first video I posted above. The 2 carriages are therefore quite far apart and would be impractical to wire together, not to mention putting 2 motors on one hub would lose the lighting function for that train carriage. What I want to achieve is:

-On the lego remote, if I press and hold left + the train will go forward at power 50%. If i release the button, the train will stop.

-If I press and hold left minus, the train will go backward at power minus 50%. If I release the button, the train will stop.

-If I press right +, the train power will increase from whatever power it was, (as per my code, usually initial power 0) by 10%. I can release the button and train will maintain the current power percentage in a forward direction.

-I I press right minus, the exact same as per my prior bullet point but reversed. Note that this means you can slow down the train and then eventually reverse it from a forward running position. Note also that I have input small wait intervals for the the right side operations such that pressing the button does not accidentally register multiple presses and shoot up to 100% in just one quick press.

-Pressing either red button, 'right' or 'left' will stop the train dead at 0% power.

-The lights on both carriages go on once the first button is pressed and they stay on until the hub is powered off.

-The operations are broadcast from the main hub to the receiving hub and the receiving hub simply copies the main hub (but mulitplies by minus 1 so that it is going backwards rather than forward, which makes sense when you see the first video I posted; it is working 'with' the main motor rather than pulling against it).



I tried modifying the code with tasks so it's a little cleaner and also by incorporating your 'wait until not' rather than using a repeat loop as I have a suspicion that the repeat loops might be causing the stuttering. Unfortunately, it makes no difference and I still have the lag and worse yet, total stops for up to 3 or 4 seconds at a time on the receiving (rear motor) hub. The main hub, as before, operates fine without issue.

![image](https://github.com/pybricks/support/assets/114411808/72b9c624-a3e6-4e7a-a286-8f752fe0cd89)

The support hub remains unchanged:

![image](https://github.com/pybricks/support/assets/114411808/7980cd09-79b8-4220-bd4f-5ce3edfd9e00)



I will attempt your exact code above and see if it is as responsive as your video; I'm hoping it will be and that this is just a case of refining the code.

(I'm wondering if it's possible to achieve what I want to without the 'if' conditions and instead have some sort of wait until button pressed, then see what it is, then perform task then go back to the beginning as opposed to the chain I have now. Kinda like reworking the code in parallel rather than series so that fewer operations run at a time rather than sort of checking through all of them. I suspect I'm talking rubbish though as the lags should not be so pronounced but maybe....)

On the remote, based on this video: https://www.youtube.com/watch?v=2Ug9c0eeWx0 and others similar, it looks as though a remote can connect to multiple city hubs using standard firmware. I tried mimicing the steps with Pybricks firmware installed but cannot successfully get it to connect. When you mentioned that this is a limitation of the firmware, did you mean that this functionality has not carried over from the Lego firmware to Pybricks?

**Reply by laurensvalk** [reply]
Thanks for the details. It may be a little while before we make such an example, but we'll put it on our list!

If the minimal delay in the video above is satisfactory, I think we can make this work.

> it looks as though a remote can connect to multiple city hubs using standard firmware

In those cases, I think the remote also connects to just one hub, and the hubs bridge the signal to the other hubs. I'd be happy to be proven wrong though.

**Reply by EmperorKabir** [reply]
Thank you! It will be great to see a clean version.

For the record I have now got it sort of working but not ideal. I'm using the above code but I've set the wait on the main program at the end to 150ms but the main thing is that I have added a wait of 500ms to the receiving hub. It seems to suggest that the receiving hub is hardware limited in how quickly it can receive and unpack the bluetooth from the main hub. Any faster than this and lags/stutters come into play. I'm unsure if this was something you already knew but it looks as though for rapid response between hubs, 500ms apperas to be the approximate limit between polling/processing the bluetooth and command.

Edit: While connected to pybricks, the limit seems to be about 500 but when disconnected from Pybricks, this can be brought down to about 300ms. (I've set it to 330 as a buffer).

**Reply by laurensvalk** [reply]
I am going to cheat a little and introduce a new feature to simplify this type of program. :smile: 

A common thing to do is to wait for *any* button to be pressed (in a wait block), and then decide what to do based on *which* button was pressed (with an if/else block).

We can already do this in Python, but it's a bit cumbersome in blocks for the moment.

So I think it makes sense to introduce an extra option in the button list that says "any button" is pressed, or something like that.

Then your program can look like this. Repeat forever:

- Wait for any button to be pressed.
- Depending on the button, increase/decrease/reset speed.
- Broadcast the new speed, and turn on the one motor at that speed too.
- Wait for all buttons to be released (that is, wait until *not* any button is pressed).

This would also eliminate the buffer wait you've mentioned, since we only have to wait for exactly as long as the buttons are pressed.

> It seems to suggest that the receiving hub is hardware limited in how quickly it can receive and unpack the bluetooth from the main hub. Any faster than this and lags/stutters come into play.

The receiving aspect happens in the background, so it should be mostly instantaneous. I'll have to have a look at this.

**Reply by EmperorKabir** [reply]
Oh that would be brilliant! I'll be very keen to test this!

**Reply by laurensvalk** [reply]
Here's what it might look like. I think you'll also like the reduced number of options, showing only the buttons available on that device.

![image](https://github.com/pybricks/support/assets/12326241/9ce319ac-3f9f-4959-8845-ddbf826f5de9)

**Reply by EmperorKabir** [reply]
That's what I thought but what block do you use afterwards as presumably you'd need to go back into an if statement if trying to distinguish different buttons causing different actions? In this form, it would only be useful to perform a singular action upon any button press rather than directly work for the type of code I was attempting?

**Reply by laurensvalk** [reply]
> That's what I thought but what block do you use afterwards as presumably you'd need to go back into an if statement if trying to distinguish different buttons causing different actions?

Yes. See the "pseudo code" in the bullet list above.

>  In this form, it would only be useful to perform a singular action upon any button press rather than directly work for the type of code I was attempting?

It's a step towards a way that I think your code would work well. Closer to your context, it's something like this.

![image](https://github.com/pybricks/support/assets/12326241/f1082507-c1b5-467d-b0ef-3050452bc2a3)

It's good timing because I'm working on the buttons for another upcoming feature... So hopefully we can release this functionality as a beta version in the coming week.

Stay tuned!

**Reply by laurensvalk** [reply]
While we wait on the next release, here is a first step that I think has most of the requirements you listed above, to serve as some inspiration for the time being.

Note that this program does not broadcast anything yet or turn on any motors. Instead, it uses the `print` block to show what it *will* send in the next version of the code.

This can be a useful way to test your logic before making things move or transmit over the air. You can see the numbers at the bottom.

![image](https://github.com/pybricks/support/assets/12326241/0eefc0a4-903a-4167-a438-8079ffb46ade)

**Reply by EmperorKabir** [reply]
Thank you, though why does this run more efficiently/quickly than the same code but with removing the 'wait until any button is pressed' line?

**Reply by laurensvalk** [reply]
The efficiency is the same, but I think this eliminates a few of the other waits/loops in your code.

I am getting this result on a train. There is some delay as expected but not too bad. What do you think?

In the second half of the video I use another two-port hub from SPIKE Essential. (It could just as well fit in a train, I just don't have that many trains). Its Bluetooth chip is a bit better, so if it is the sender, there are even less delays.

https://github.com/pybricks/support/assets/12326241/09eee0ec-e0fa-47ff-a4eb-2fab77d55a40

**Reply by EmperorKabir** [reply]
I must admit that certainly looks slightly faster and there appear to be no stutters. Did you try holding it down for a longer period of time? Eg 10-20 seconds? And you have not put a wait delay on the receiving hub?

**Reply by laurensvalk** [reply]
I just ran it again for 30+ seconds without stutter. I did not put in any wait delays other than waiting for any button to be pressed and then released, and no waits on the receiving end. It shouldn't hurt to add one though.

**Reply by EmperorKabir** [reply]
Excellent! I'll await the button release and will give it a try :)

**Reply by laurensvalk** [reply]
Would you look at this :smile: 

Some more work in the firmware....

@dlech , for context, this is with the `ADV_INT` left at `40` everywhere as per the email.

https://github.com/pybricks/support/assets/12326241/644488aa-4bce-4824-a035-ffa525c5a509

**Reply by EmperorKabir** [reply]
Exciting! So does this mean I'm awaiting a new firmware update and additional block(s)?

**Reply by laurensvalk** [reply]
Yep! Updating everything at once...

**Reply by laurensvalk** [reply]
We released a new beta version. You can try it out with the new firmware at https://beta.pybricks.com/

If you've used that page before, it may take a few minutes for the latest version to load in the background. If you go to settings/about, you'll see 3.4.0b1 as the new version. When you do, you can go ahead and update the firmware or try new features.

![image](https://github.com/pybricks/support/assets/12326241/f7a38ea8-df61-4737-a4e6-0424bd4b3876)

**Reply by EmperorKabir** [reply]
I can see the wait until any button is pressed command :)

I tried to look at the changelog but it doesn't appear to be updated.

Am I correct in thinking that this update has added the aforementioned button as well as removed the necessity to add a wait interval on a receiving hub?

e.g. before: sending hub sets motor speed
receiving hub unpacks motor speed and sets its own motor speed but requires a 330ms wait before repeating to avoid stutters/delays

after: as above but remove the wait interval and all should be fine?

(I've yet to test this but will soon :) )

Are there any other noteable changes e.g. the killswitch block?

**Reply by laurensvalk** [reply]
Yes. In the next couple of days I'll post the example used for the videos. Feel free to remind me here if I haven't.

Indeed, you've already found the blocks, and you can update the firmware as usual to get the latest there too. City hub broadcasting and observing should be nearly instant now, and not crash. 

The changelog and various code updates will be released in a few weeks time.

This is an early release to get some testing on some improvements under the hood :)

**Reply by EmperorKabir** [reply]
So I tried testing this code using the new firmware (and updated the beta app):

Sending hub:
![image](https://github.com/pybricks/support/assets/114411808/de7fc6bd-7798-493d-90c2-6c5577a2ec71)

Receiving hub:
![image](https://github.com/pybricks/support/assets/114411808/179eb47d-5a18-43ef-8dc0-fa5df7ad3081)

As you can see, I have removed the 330ms wait from the receiving hub.

Here is the result. Note that the hubs are not connected to the computer in this testing:

https://youtu.be/VwlNBDGlCfw

and here is the result when I add the 330ms wait back in:

https://youtu.be/HS6QnE8qGrk

On the previous firmware, the 330ms wait was a halfway house workaround for it to kinda work on a delay. With the updated firmware, the lag is still present but additionally, I'm not sure the 330ms delay halfway house workaround is working like it was before.

I tested this again with my work in progress code for the sender hub which starts, accelerates progressively, hits a top speed, begins progressive deceleration when it hits a tile and then stops at a station, then repeats. There appears to be little difference and the lag remains similar:

![image](https://github.com/pybricks/support/assets/114411808/e94873b7-fabf-4697-bbad-4c3142045e70)

Strangely with this more complex code, the differences appears negligble and variable in terms of which one is better!

I hope this feedback is of use!

(Can I please also ask where the block is to turn off / power down the hub?)

(I also wanted to check why adding the 'wait until any button on the remote was pressed' before the if statement makes the code more efficient when it is just adding an extra line which intuitively to me seems like adding 'bloat' code? The code in which I have added the new function is here:)

![image](https://github.com/pybricks/support/assets/114411808/cc3b5568-383f-4206-9f48-f9d0915fd30d)


Apologies for talking about 3 different things! I'm more than happy to put them in new threads if this would be preferred!

And thank you again!

**Reply by laurensvalk** [reply]
Isn't the motor on the sender just stopping because it sets the duty cycle to 0? Or was that intentional?

Did you update the firmware on both hubs? You can see the installed version at the bottom right of the app when the hub is connected.

> Can I please also ask where the block is to turn off / power down the hub?

The hub poweroff block is not included yet. Some other ideas: a program stop block. Less abrupt, and then it will auto-turn off in three minutes anyway. See https://github.com/pybricks/support/issues/1405

> I also wanted to check why adding the 'wait until any button on the remote was pressed' before the if statement makes the code more efficient when it is just adding an extra line which intuitively to me seems like adding 'bloat' code? 

It's just one way to intuitively group all waiting operations together. If I have just two wait blocks in my entire program, I know that's where the only "delays" are allowed to be, and they are also easily described based on something you can see ("waiting on any button").

> seems like adding 'bloat' code? 

It ensures that the loop goes around only once when there is something new to do, i.e. change the send signal only when the button state *changes*. This is probably also less likely to confuse the Bluetooth chip, since it isn't resetting the same signal over and over by running the broadcast block thousands of times per second.

> Apologies for talking about 3 different things! I'm more than happy to put them in new threads if this would be preferred!

If you're finding that I miss certain responses, then probably, otherwise no worries :smile: 

I think you're getting the hang of this pretty quick -- wireless communication and the logic to deal with it is usually only the last chapter of a typical programming book, so you'll be just fine :smile:

**Reply by EmperorKabir** [reply]
Thank you, I'll reapproach the code with the remote and see if I can minimise the broadcasting.

Yes, the hubs both have their firmware updated. The sender hub is stopping periodically on purpose so that I can check if the receiving hub is fast enough to also stop periodically and also get a visual cue on the lag. In both videos, it is not fast enough to ever stop.

**Comment by laurensvalk** [comment]
I have added one possible way of doing it in https://github.com/pybricks/pybricks-projects/issues/96. We might eventually add it to the site, but it's good to have a place for it now.

**Reply by EmperorKabir** [reply]
Ahhh that's starting to make sense! I am attempting to mimic this code and test it but have 2 questions. I have added the variable (where you have named it power level and I named mine power percentage) but am unable to find it as a variable in the blocks to add to the list, (mine is showing as list with 0 at the moment). I tried bringing in the bottom one but that told me to set up device first.

I also wondered whether or not the receiving hub would need to amend it's unpacking to power level in yours rather than just power or whether the name doesn't matter. (If the name does not matter, how does it distinguish if it was unpacking multiple variables? Would it be the order of the variables?)

![image](https://github.com/pybricks/support/assets/114411808/9d05f8f4-06b0-4313-8475-b1bad4331ef1)

**Reply by laurensvalk** [reply]
The names don't matter. You're right, it would be better to make it consistent.

Indeed, in a list the order is what matters. We could also update the example to use no list at all, which would be even simpler. Maybe in a few days. Thanks!

**Reply by laurensvalk** [reply]
As for those yellow round blocks, they are the function arguments. Once you've added one to your task (by expanding the white arrow), you can just pull out the argument from there.

**Reply by EmperorKabir** [reply]
I tried this code and the stutters/lags are about as pronounced as before like in the previous 2 youtube videos I posted. Is there any correction I can make to the waits? (I have very similar results when I remove the 66ms and 330 ms waits as well)

![image](https://github.com/pybricks/support/assets/114411808/b2a7dff1-36f4-414e-b817-497df5c04e51)
![image](https://github.com/pybricks/support/assets/114411808/f8d0c9de-f66b-42c6-a354-89eedf9cacd6)

https://youtu.be/Kb47yZabMn0

**Reply by laurensvalk** [reply]
I don't seem to see the block that waits for the buttons to be released. Could you try the exact example I've shared to see if it eliminates the delays?

Just to rule out any other issues, ceteris paribus.

Also be sure to disconnect the receiving hub from the computer.

I've found that the sender hub still operates smoothly while connected, which is really neat because this is probably where you spend the most time tinkering with the code anyway. (With the receiver code hardly changing once you've got it right).

**Reply by EmperorKabir** [reply]
I just tried your code. The lag is a little more pronounced than in your video but somewhat acceptable (typically around 0.5 seconds lag and at the extreme end, 0.75ish seconds). -edit- Just tried with it disconnected from the computer and it is much more in-line with your videos at approximately 0.1-0.2 seconds leg (All these numbers are by eye guesses lol). Looks like the firmware/code is working well, thank you! :)

(This in turn made me wonder if it's worth submitting an idea of putting the bluetooth connections with the PC in some sort of suspended/idle/low use mode while the program is running so that you don't need to disconnect each time for real world testing and then when the program stops, it just ramps back up again - perhaps minimise broadcasting/receiving from PC while program is running? - however I have next to no knowledge on how bluetooth works to even know if this is even something remotely possible).

Is the nested if loop within an if loop by intention? i.e. more efficient than just elseif? I suppose I need to figure out where to put the block within my own code that is suffering from lag/stutter that waits for buttons to be released. I'm also hoping that additional code (.e.g turning lights on and looking at colour sensor data/actions), would not indirectly increase the lag; I'm guessing not as I suspect larger code is nowhere near as big of a deal as efficiently coding the broadcasting lines.

I have a strange seperate issue where the sending hubs wheels will not turn at the very low speeds, e.g. 10% power but the receiving hub does. I suspect this might simply be battery drain but I just wondered if there's any outside chance that additional power drain on the sending hub with more code on it and the power required to broadcast would cause 'not enough current/voltage' (my electrical knowledge sucks) compared to the receiving hub which is somehow using less and is OK. I'm guessing it's just the former though.

**Reply by laurensvalk** [reply]
Thanks for the detailed report. Glad it's working!

There is one thing we may be able to reduce while connected, but I don't know yet if this will improve observing.

However as mentioned earlier, depending on your setup, you may not need to change the receiving code much, so it can stay disconnected most of the time anyway.

The choice of the if blocks isn't about efficiency here but about making the code easier to follow.

The first if decides whether we're going to be in on/off mode, and the other devices that, if not, how to change the current speed.

Regarding the 10% speed, that's just about enough (or not) to overcome the inner friction, which varies between motors. Maybe the other will start at 12%. The best thing would be to adapt the program to just skip 10% altogether because it's probably not enough to pull carriages anyway.

**Comment by EmperorKabir** [comment, marked as answer]
Thank you so much for all your responses in this thread; it's very much appreciated, especially your patience! :)

I finally got a slightly modified version of your code working here:

![image](https://github.com/pybricks/support/assets/114411808/b01f22ff-c79e-4443-ab6e-757244f6a45b)

The receiver is relatively unchanged.

Note: I have 2 pairs of hubs and motors (I have clearly spent too much money on Lego trains....) and I noticed that I have the same issue with the motor not having enough power to run at the 10% level on the sending hub but it does move the wheels on the receiving hub. This could be conincidence but it certainly looks like some sort of additional power reduction based on what the sending hub is doing causing the motor to lack enough actual power to turn. As you say, this is perhaps something that cannot be helped, though without speed testing equipment, I would otherwise have built in a power difference between the sending and receiving hub. Again though, this is likely too negligible to matter... for trains at least! Just thought it might be something you'd like to know. e.g. if there is a function to read back the motor speed and adjust that accordingly rather than simply outputting a power level.

**Reply by laurensvalk** [reply]
You can think of the duty cycle as the percentage of the available battery level applied to the motor.

So it's quite possible you've used the sending hubs just slightly more since your last charge.

**Reply by EmperorKabir** [reply]
So will presumably having lights/colour sensor plugged in, immediately causing a little trickle drain even if unused. Possibly a power difference between broadcasting and receiving and code processing too I guess. Thank you!
